runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/WindmillStubFactoryFactoryImpl.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.client.grpc.stubs;
+
+import static org.apache.beam.runners.dataflow.worker.windmill.client.grpc.stubs.WindmillChannelFactory.remoteChannel;
+
+import com.google.auth.Credentials;
+import java.util.function.Function;
+import org.apache.beam.runners.dataflow.options.DataflowWorkerHarnessOptions;
+import org.apache.beam.runners.dataflow.worker.windmill.WindmillServiceAddress;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+
+public class WindmillStubFactoryFactoryImpl implements WindmillStubFactoryFactory {
+
+  private final int windmillServiceRpcChannelAliveTimeoutSec;
+  private final Credentials gcpCredential;
+
+  public WindmillStubFactoryFactoryImpl(DataflowWorkerHarnessOptions workerOptions) {
+    this.gcpCredential = workerOptions.getGcpCredential();
+    this.windmillServiceRpcChannelAliveTimeoutSec =
+        workerOptions.getWindmillServiceRpcChannelAliveTimeoutSec();
+  }
+
+  @Override
+  public WindmillStubFactory makeWindmillStubFactory(boolean useIsolatedChannels) {
+    Function<WindmillServiceAddress, ManagedChannel> channelFactory =
+        serviceAddress -> remoteChannel(serviceAddress, windmillServiceRpcChannelAliveTimeoutSec);
+    ChannelCache channelCache =
+        ChannelCache.create(
+            serviceAddress ->
+                // IsolationChannel will create and manage separate RPC channels to the same
+                // serviceAddress via calling the channelFactory, else just directly return the
+                // RPC channel.
+                useIsolatedChannels
```

Comment:
would it be simpler to just have IsolatedChannel accept a atomic boolean that it then examines on whether or not it should perform the isolation or not?

It seems like we wouldn't have to have as much of the plumbing or the factoryfactory if we just always created an IsolatedChannel and just had a way to disable it's effect.

Reply:
The channels are cached by ChannelCache, so we have to somehow invalidate the caches when the setting changes. Current approach creates a second new ChannelCache and let the old one expire. Recreating the cache seems simpler than having to invalidate channels, track usage and expire them individually.

Reply:
What I was describing wouldn't involve destroying the channels, we'd always be using IsolationChannels they would just have some option to turn off isolation and use a single channel for all calls internally (alternatively we could add some max channels to IsolationChannel and set it to 1 dynamically for this).

So we wouldn't have to invalidate caches, but we would have to have some way to get the desired change to all of the channels.  You could have all isolation channels register for interest in changed global configs when created and then individually they modify themselves for example.

But since this is done, doesn't seem too bad and we are likely to remove the flag after rollout let's just go with this.

Reply:
Ack.